### ISSUE TITLE
Loading in a model with Prioritized Experience Replay causes uninitialized model?

### ISSUE THREAD

**Issue by jandevries123** [hill-a/stable-baselines#940]
Hello, 

I found that when i load in my model as shown in the examples in the documentation that my model starts of uninitilized.

I assume this out of:

Probability [0.34276846 0.31248772 0.34474385]
Probability [0.3427647  0.312495   0.34474024]
Probability [0.3427545  0.31251618 0.3447293 ]
Probability [0.3427277  0.3125703  0.34470204]
Probability [0.34271523 0.31259626 0.34468853]

Example above illustrates the output of a DQN policy, this behaviour occurs throughout the entire run or occurs in small segments of the run.

I was thinking this was due to my custom environment being buggy, however i debugged that everything is working accrodingly as it does when training the model. When training the model, nothing goes wrong and the environment just works, however in testing it goes wrong. I have sessions where the trained model just works when loaded, but i also have results where this comes out, the above. I highly assume the model is loaded in uninitilized as i know from experience that the distribution from the start of a trained model is ca [0.4, 0.2, 0.4], usually. With this in mind i can assume there is something wrong. After debugging and finding that my environment seems to just work fine as it is literally just receiving these distributions from the network and using it, i am assuming something is wrong. The action 3, aka the always chosen action represent the action "Do Nothing" where the agent does nothing, i have a terminal steplimit at a 1000 steps, meaning i get a run where the agent does nothing for episodes long. This occured in both LnMlpPolicy and MlpPolicy

I am not 100% where this issue comes from as it comes in 2 forms, either the entire run like this or small segments like this, but since when i am training the model it all seems to work just fine and debugging, i am not sure what or where this bug is coming from as doing nothing gives a 0 reward input and the distribution of the output also do not look OK, However i do thing that atleast part of it comes from the prioritized experience replay, because when i use this it most of the times causes this issue, however i also had sessions where this was still the issue without prioritized experience replay (segments). I get that this would be difficult especially with my custom environment, but again, the training works exactly as intended.

Is there any known issue or thing i am missing out on? It has been "bugging" me all nigh(s) long.

Settings:
`{
    "double_q": true,
    "param_noise": false,
    "learning_starts": 1000,
    "train_freq": 1,
    "prioritized_replay": true,
    "prioritized_replay_eps": 1e-06,
    "batch_size": 256,
    "target_network_update_freq": 500,
    "prioritized_replay_alpha": 0.6,
    "prioritized_replay_beta0": 0.4,
    "prioritized_replay_beta_iters": null,
    "exploration_final_eps": 0.02,
    "exploration_fraction": 0.1,
    "learning_rate": 0.0005,
    "gamma": 0.99,
    "verbose": 0,
    "observation_space": {
        ":type:": "<class 'gym.spaces.box.Box'>",`

**Comment by Miffyli** [comment]
Please include a minimal example code to replicate the issue.

Loading models with `load` should not leave the "uninitialized", they should match the behaviour they had when they were saved. Make sure you use `deterministic=True` in the `predict` function.

**Comment by jandevries123** [comment]
> Please include a minimal example code to replicate the issue.
> 
> Loading models with `load` should not leave the "uninitialized", they should match the behaviour they had when they were saved. Make sure you use `deterministic=True` in the `predict` function.
> 
> Edit: Replay memorys are not included in the saved models to reduce space usage on disk.

I will look into making a minimal code example later, but replay memorys that aren't included in the saved models shouldn't cause this behaviour?

**Comment by Miffyli** [comment]
>  but replay memorys that aren't included in the saved models shouldn't cause this behaviour?

Nope, unless you continue training the model, in which case you have lost the previous data you had.

> and why is deterministic=True so important here?

Seems like the issue is on "getting different actions than expected", and `deterministic=True` makes sure you will get same actions for same states.

**Comment by Miffyli** [comment]
Hmm indeed the code should work as expected. I would still check the performance of the agent before saving and after loading it by playing fixed number of episodes and compare those. Another way to check it is to check if model parameters match before saving and after loading with `model.get_parameters()`.

**Comment by jandevries123** [comment]
> Hmm indeed the code should work as expected. I would still check the performance of the agent before saving and after loading it by playing fixed number of episodes and compare those. Another way to check it is to check if model parameters match before saving and after loading with `model.get_parameters()`.

I have a very trustworthy theory about it. I think it is because of the RAM in Google Colab clogging up. I don't know why i haven't seen this happen earlier, but it makes sense. If there isn't enough RAM he won't be able to load in the model properly, atleast that is what i am thinking...

**Comment by jandevries123** [comment]
> I do not think that is the cause, as memory issues would usually result in fatal crash, unless Colab does something odd. Again, a performance comparison of policies would be a confirmation that something is off.

Yes i realized this would be a strange case if it were to be so, just after i sent it. I will check model.get_parameters() now. What did u mean with testing diffrent policies? like if the same issue occurs on diffrent polices?

**Comment by Miffyli** [comment]
By "testing" I refer to running the agents on the environment and collecting episodic rewards (sum of rewards per episode) and comparing the average result between policies. See [evaluate_policy](https://stable-baselines.readthedocs.io/en/master/common/evaluation.html).

**Comment by jandevries123** [comment]
Hello,

After several tests i discovered that this issue only occurs when loading in a model which has leaky_relu as activation function, and when i say only, i mean only when loading in this activiation function. I don't know why though.

**Comment by Miffyli** [comment]
@jandevries123 

Could you provide a code to replicate the issue? Sounds like something is not stored/loaded related to activation functions, but as far as I know leaky ReLUs do not have any trainable parameters (only the "leakyness" parameter, which could be trained).

**Comment by jandevries123** [comment]
I am really not using any special code, i think.

`for i in range(len(filenames)):
  print("Evaluating Model", i, "with name", filenames[i])
  if filenames[i].startswith('DQN'):
    model = DQN.load(filenames[i])  
    env = gym.make("MyEnv", use_lstm=False)
    obs = env.reset()
    for i in range(1049307)::
        action, _states = model.predict(obs)
        obs, rewards, dones, info = env.step(action)
        if not _ % 10000:
            print(_)`

**Comment by Miffyli** [comment]
I mean full code to replicate the issue, i.e. the custom policy, training of the model, saving of the model, loading the model and testing it out. A piece of code that, when ran with correct with dependencies installed, should raise up the issue.

**Comment by jandevries1421353151** [comment]
jan from another dimension here.

So i found that the unintialized loading problem comes from when i put deterministric to true, aka default when i do i get probabilities that look very unitialized like ca[0.33, 0.33, 0.33] when if i were to do deterministic to false i get the desired values i was looking for like [0.44, 0.1, 0.46], but deterministic causes random actions of course so it is not functional, how comes it does work normal on deterministic false?

**Comment by Miffyli** [comment]
I do not see how the `deterministic` flag could affect the initial parameters. Again, we need the full code to replicate the issue (i.e. "this code should do X but instead it does Y") to be able to tell what could be wrong.

**Comment by jandevries1421353151** [comment]
For training i use:

```
                model = DQN(policy, Env, verbose=verbose, tensorboard_log=("tensorboard/DQN " + str(runs) + "/"), prioritized_replay=prior, batch_size=batch_size, learning_rate=learning_rate, gamma=gamma, policy_kwargs=policy_kwargs, double_q=double_q, train_freq=train_freq)
                model.learn(total_timesteps=timesteps, tb_log_name=run_name)
                model.save(run_name)
```

then for loading the model i use:"

```
    model = DQN.load(filenames[i])  
    obs = env.reset()
    for i in range(23000000):
        action, _states = model.predict(obs, deterministic=True)
#        print(model.action_probability(obs, actions=None))
        obs, rewards, dones, info = env.step(action)
```

**Comment by Miffyli** [comment]
For completeness and to avoid any confusion, please include one complete code to replicate the issue, including printing out the probabilities of the actions which you have used to debug the issue.

**Comment by jandevries1421353151** [comment]
Hello,

sorry for the late response. Here are images of the probability distributions of the output of "bugged" and "working" trained sessions:

![2020-10-23_18h28_53](https://user-images.githubusercontent.com/40467919/97078641-cd41b500-15ed-11eb-90d3-86de290aa0d5.png)
![2020-10-23_18h28_25](https://user-images.githubusercontent.com/40467919/97078649-d599f000-15ed-11eb-9fcc-771d042e942a.png)
![2020-10-23_17h13_37](https://user-images.githubusercontent.com/40467919/97078650-d7fc4a00-15ed-11eb-84d4-8b102d780933.png)
![2020-10-23_17h05_31](https://user-images.githubusercontent.com/40467919/97078654-da5ea400-15ed-11eb-85b4-682a3148aff8.png)
![2020-10-23_17h01_07](https://user-images.githubusercontent.com/40467919/97078656-dc286780-15ed-11eb-88f0-c8d6f7bd5bfb.png)

As you can see, weird gaps that causes the agent to do nothing. While the could be a coincidence it is far from expected behavior from the agent since doing nothing the entire episode long gives 0 reward. It appears to probably be a bug, however i debugged the state, reward, terminal and all that stuff and nothing seems to be out of the ordinary except the probabilities themselves. I also found out that it is not settings related, it just happens from time to time almost seemingly randomly. How would i further debug this issue and what is causing it? I think it is loading in the model incorrect perhaps? What are your thoughts?

**Comment by Miffyli** [comment]
Without knowing more about the environment and other settings it is hard to say if this is expected behaviour or not, but generally you can see sudden dips in performance during training. **Again**, if you believe this is a bug or an issue in stable-baselines, please share the exact and full code to replicate the issue. Otherwise you can close this issue. Do note that we do not offer technical support.

**Comment by araffin** [comment]
@hill-a , as requested by @jandevries123 (by email), could you delete this issue? (only the owner of the repo can apparently...)
